### ISSUE TITLE
Incorrect expression is used to represent the smallest integer in flang module

### ISSUE THREAD

**Issue by chilikink** [llvm/llvm-project#55086]
Compilation of the following code
```
USE, INTRINSIC :: ISO_FORTRAN_ENV
COMPLEX(REAL32), PARAMETER :: C1 = (+2147483647_INT32, -2147483648_INT32)
COMPLEX(REAL32), PARAMETER :: C2 = (+2147483648_INT32, -2147483649_INT32)
INTEGER(INT32), PARAMETER :: I1 = -2147483647_INT32
INTEGER(INT32), PARAMETER :: I2 = -2147483648_INT32
INTEGER(INT32), PARAMETER :: I3 = +2147483647_INT32
INTEGER(INT32), PARAMETER :: I4 = +2147483648_INT32
END
```
with flang on the branch release/14.x (commit 9a3e81e1f91fa8793857fdfdacfd7e70ac365bb8) results in the following errors:
```
$ flang -c test.f90
error: Semantic errors in test.f90
./test.f90:3:37: error: Integer literal is too large for INTEGER(KIND=4)
  COMPLEX(REAL32), PARAMETER :: C2 = (+2147483648_INT32, -2147483649_INT32)
                                      ^^^^^^^^^^^
./test.f90:3:56: error: Integer literal is too large for INTEGER(KIND=4)
  COMPLEX(REAL32), PARAMETER :: C2 = (+2147483648_INT32, -2147483649_INT32)
                                                         ^^^^^^^^^^^
./test.f90:5:36: error: Integer literal is too large for INTEGER(KIND=4)
  INTEGER(INT32), PARAMETER :: I2 = -2147483648_INT32
                                     ^^^^^^^^^^
./test.f90:7:36: error: Integer literal is too large for INTEGER(KIND=4)
  INTEGER(INT32), PARAMETER :: I4 = +2147483648_INT32
                                     ^^^^^^^^^^
```
The errors are not consistent. For complex parameters, the minimum integer of the type `INT32` is `-2147483648`, while for integer parameters the minimum integer of the type `INT32` is `-2147483647`. The maximum value is the same in both cases (2147483647). The integer literal boundaries are tested in the file `flang/test/Semantics/int-literals.f90`, but the test is only performed for the complex-type parameters there.

**Comment by klausler** [comment]
It's due to a subtle point of Fortran syntax that -2147483648_4 is not a constant that's in range.  The F'2018 standard's grammar requires that "constant" to be parsed as an "int-literal-constant" (R708) that is an operand to a "level-2-expr" (R1006)., namely the minus sign.  So it parses as `-(2147483648_4)` and the range check applies *before* the negation..

In other words, Fortran's grammar only has positive integer constant literals as expression primaries.  This goes way back to an incompatible change in Fortran 90; FORTRAN '77's grammar had signed integer literal constants (pg. 4-3).

The portable ways to write that parameter initialization expression, which also avoid the analogous error message from GNU Fortran, is `-2147483647-1` or `-HUGE(0_INT32)-1`.

**Comment by chilikink** [comment]
Sorry, I reduced the original error too much. Thank you for the explanation. Here is an updated test case
```
$ cat test.f90
MODULE M
  USE, INTRINSIC :: ISO_FORTRAN_ENV
  INTEGER(INT32), PARAMETER :: I = -2147483647-1
END
$ cat test2.f90
USE M
END
$ flang -c test.f90
$ flang -c test2.f90
error: Semantic errors in test2.f90
./m.f18.mod:48:26: error: Integer literal is too large for INTEGER(KIND=4)
  integer(4),parameter::i=-2147483648_4
                           ^^^^^^^^^^
```
where flang itself uses the incorrect syntax in module definition.

**Comment by klausler** [comment]
Fixed by D126584.
